src/reference/mod.rs

```diff
@@ -0,0 +1,367 @@
+//!     double-colon: `@node::variable`.
+//!
+//! Local and scoped variables are created using `var` or `let` statements.  A `let` statement
+//! creates an **_immutable variable_**, whose value cannot be changed.  A `var` statement creates
+//! a **_mutable variable_**.  You use a `set` statement to change the value of a mutable variable.
+//!
+//! (All global variables are immutable, and cannot be created by any graph DSL statement; they are
+//! only provided by the external process the executes the graph DSL file.  If you need to create
+//! your own "global" variable from within the graph DSL, create a scoped variable on the root node
+//! of the syntax tree.)
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   let local_variable = "a string"
+//!   ; The following would be an error, since `let` variables are immutable:
+//!   ; set local_variable = "a new value"
+//!
+//!   ; The following is also an error, since you can't mutate a variable that
+//!   ; doesn't exist:
+//!   ; set missing_variable = 42
+//!
+//!   var mutable_variable = "first value"
+//!   attr @id.node "first_attribute" = mutable_variable
+//!
+//!   set mutable_variable = "second value"
+//!   attr @id.node "second_attribute" = mutable_variable
+//!
+//!   var @id::kind = "id"
+//! }
+//! ```
+//!
+//! Variables can be referenced anywhere that you can provide an expression.  It's an error if you
+//! try to reference a variable that hasn't been defined yet.  (Remember that stanzas are processed
+//! in the order they appear in the file, and each stanza's matches are processed in the order they
+//! appear in the syntax tree.)
+//!
+//! # Regular expressions
+//!
+//! You can use a `scan` statement to match the content of a string value against a set of regular
+//! expressions.
+//!
+//! Starting at the beginning of the string, we determine the first character where one of the
+//! regular expressions matches.  If more than one regular expression matches at this earliest
+//! character, we use the regular expression that appears first in the `scan` statement.  We then
+//! execute the statements in the "winning" regular expression's block.
+//!
+//! After executing the matching block, we try to match all of the regular expressions again,
+//! starting _after_ the text that was just matched.  We continue this process, applying the
+//! earliest matching regular expression in each iteration, until we have exhausted the entire
+//! string, or none of the regular expressions match.
+//!
+//! Within each regular expression's block, you can use `$0`, `$1`, etc., to refer to any capture
+//! groups in the regular expression.
+//!
+//! For example, if `filepath` is a global variable containing the path of a Python source file,
+//! you could use the following `scan` statement to construct graph nodes for the name of the
+//! module defined in the file:
+//!
+//! ``` tsg
+//! (module) @mod
+//! {
+//!   var current = @mod.root
```

Comment:
Does `root` have special significance in this DSL, or is this just a tag?

Reply:
Just a tag name like any other